src/shogun/classifier/Perceptron.cpp

```diff
@@ -65,15 +67,18 @@ bool CPerceptron::train_machine(CFeatures* data)
 	ASSERT(num_vec==train_labels.vlen)
 	SGVector<float64_t> output(num_vec);
 
-	SGVector<float64_t> w = get_w();
+	auto w = SGVector<float64_t>(num_feat);
```

Comment:
The allocation should be moved into the if

Reply:
i.e. just define it here

Reply:
actually, why dont you just do m_w.resize(num_feats)?

Reply:
m_w is private so we will need to use get_w() everywhere instead... this causes type errors with v.add(gradient, w)

Reply:
ok cool. But still you can move the allocation inside the if, you dont want to allocate memory if the user has provided intitialization